File: src/lib/repository.ts
import { supabase } from "./supabase"
import type { SupabaseClient, PostgrestError } from "@supabase/supabase-js"

/**
 * Repositorio base con CRUD genérico sobre Supabase.
 * Extiende esta clase para cada tabla y mantén el código DRY.
 *
 *   class ProductRepo extends BaseRepository<Product> {
 *     constructor() { super("products") }
 *   }
 */
export abstract class BaseRepository<T extends { id?: string }> {
  protected readonly client: SupabaseClient
  protected readonly table: string

  constructor(table: string, client: SupabaseClient = supabase) {
    this.table = table
    this.client = client
  }

  protected handle<R>(data: R | null, error: PostgrestError | null, ctx: string): R {
    if (error) {
      // eslint-disable-next-line no-console
      console.error(`[${this.table}] ${ctx}:`, error.message)
      throw new Error(error.message)
    }
    return data as R
  }

  async list(select = "*"): Promise<T[]> {
    const { data, error } = await this.client.from(this.table).select(select)
    return this.handle(data as unknown as T[], error, "list") ?? []
  }

  async get(id: string, select = "*"): Promise<T | null> {
    const { data, error } = await this.client
      .from(this.table)
      .select(select)
      .eq("id", id)
      .maybeSingle()
    return this.handle(data as unknown as T | null, error, "get")
  }

  async create(payload: Partial<T>): Promise<T> {
    const { data, error } = await this.client
      .from(this.table)
      .insert(payload as any)
      .select()
      .single()
    return this.handle(data as unknown as T, error, "create")
  }

  async update(id: string, patch: Partial<T>): Promise<T> {
    const { data, error } = await this.client
      .from(this.table)
      .update(patch as any)
      .eq("id", id)
      .select()
      .single()
    return this.handle(data as unknown as T, error, "update")
  }

  async remove(id: string): Promise<void> {
    const { error } = await this.client.from(this.table).delete().eq("id", id)
    this.handle(null, error, "remove")
  }

  /** Borrado lógico estándar (is_active=false) */
  async softDelete(id: string): Promise<void> {
    const { error } = await this.client
      .from(this.table)
      .update({ is_active: false } as any)
      .eq("id", id)
    this.handle(null, error, "softDelete")
  }
}

File: core/repository.py
"""Repositorio base genérico sobre Supabase (POO)."""

from __future__ import annotations

from typing import Any

from .db import sb


class BaseRepository:
    """CRUD genérico. Hereda y define `table` para reutilizar."""

    table: str = ""
    select: str = "*"
    order_by: str | None = "created_at"
    order_desc: bool = True

    # ─── Helpers ───
    def _q(self):
        return sb.table(self.table)

    # ─── Read ───
    def list(self, *, where: dict | None = None, limit: int | None = None) -> list[dict]:
        q = self._q().select(self.select)
        for k, v in (where or {}).items():
            q = q.eq(k, v)
        if self.order_by:
            q = q.order(self.order_by, desc=self.order_desc)
        if limit:
            q = q.limit(limit)
        return q.execute().data or []

    def get(self, id_: Any) -> dict | None:
        res = self._q().select(self.select).eq("id", id_).limit(1).execute().data
        return res[0] if res else None

    # ─── Write ───
    def create(self, payload: dict) -> dict:
        res = self._q().insert(payload).execute().data
        return (res or [{}])[0]

    def update(self, id_: Any, payload: dict) -> dict:
        res = self._q().update(payload).eq("id", id_).execute().data
        return (res or [{}])[0]

    def remove(self, id_: Any) -> None:
        self._q().delete().eq("id", id_).execute()
